File: src/PowerShell/Microsoft.WinGet.Client/PSObjects/UninstallResult.cs
// -----------------------------------------------------------------------------
// <copyright file="UninstallResult.cs" company="Microsoft Corporation">
//     Copyright (c) Microsoft Corporation. Licensed under the MIT License.
// </copyright>
// -----------------------------------------------------------------------------

namespace Microsoft.WinGet.Client.PSObjects
{
    using System;

    /// <summary>
    /// UninstallResult wrapper object for displaying to PowerShell.
    /// </summary>
    public class UninstallResult
    {
        private Management.Deployment.UninstallResult uninstallResult;

        /// <summary>
        /// Initializes a new instance of the <see cref="UninstallResult"/> class.
        /// </summary>
        /// <param name="uninstallResult">The uninstall result COM object.</param>
        public UninstallResult(Management.Deployment.UninstallResult uninstallResult)
        {
            this.uninstallResult = uninstallResult;
        }

        /// <summary>
        /// Gets the correlation data of the uninstall result.
        /// </summary>
        public string CorrelationData
        {
            get
            {
                return this.uninstallResult.CorrelationData;
            }
        }

        /// <summary>
        /// Gets the extended error code exception of the failed uninstall result.
        /// </summary>
        public Exception ExtendedErrorCode
        {
            get
            {
                return this.uninstallResult.ExtendedErrorCode;
            }
        }

        /// <summary>
        /// Gets a value indicating whether a reboot is required.
        /// </summary>
        public bool RebootRequired
        {
            get
            {
                return this.uninstallResult.RebootRequired;
            }
        }

        /// <summary>
        /// Gets the status of the uninstall.
        /// </summary>
        public string Status
        {
            get
            {
                return this.uninstallResult.Status.ToString();
            }
        }

        /// <summary>
        /// Gets the error code of an uninstall.
        /// </summary>
        public uint UninstallerErrorCode
        {
            get
            {
                return this.uninstallResult.UninstallerErrorCode;
            }
        }
    }
}

File: src/Microsoft.Management.Deployment/UninstallResult.cpp
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
#include "pch.h"
#include "UninstallResult.h"
#include "UninstallResult.g.cpp"
#include <wil\cppwinrt_wrl.h>

namespace winrt::Microsoft::Management::Deployment::implementation
{
    void UninstallResult::Initialize(
        winrt::Microsoft::Management::Deployment::UninstallResultStatus status,
        winrt::hresult extendedErrorCode,
        uint32_t uninstallerErrorCode,
        hstring const& correlationData, 
        bool rebootRequired)
    {
        m_status = status;
        m_extendedErrorCode = extendedErrorCode;
        m_uninstallerErrorCode = uninstallerErrorCode;
        m_correlationData = correlationData;
        m_rebootRequired = rebootRequired;
    }
    hstring UninstallResult::CorrelationData()
    {
        return hstring(m_correlationData);
    }
    bool UninstallResult::RebootRequired()
    {
        return m_rebootRequired;
    }
    winrt::Microsoft::Management::Deployment::UninstallResultStatus UninstallResult::Status()
    {
        return m_status;
    }
    winrt::hresult UninstallResult::ExtendedErrorCode()
    {
        return m_extendedErrorCode;
    }

    uint32_t UninstallResult::UninstallerErrorCode()
    {
        return m_uninstallerErrorCode;
    }
}
